content/courses/Ada_For_The_C_Embedded_Developer/chapters/10_Conclusion.rst

```diff
@@ -3,6 +3,175 @@ Conclusion
+shifts, they should be relatively rarely needed. When translating C code to
+Ada, it's good practice to consider alternatives. In a lot of cases, these
+operations are used to insert several pieces of data into a larger structure.
+In Ada, this can be done by describing the structure layout at the type level
+through representation clauses, and then accessing this structure as any other.
+For example, we can interpret an arbitrary data type as a bit-field and perform
+low-level operations on it.
+
+Because Ada is a strongly typed language, it doesn't define any implicit type
+conversions like C. If we try to compile Ada code that contains type
+mismatches, we'll get a compilation error. Because the compiler prevents
+mixing variables of different types without explicit type conversion, we can't
+accidentally end up in a situation where we assume something will happen
+implicitly when, in fact, our assumption is incorrect. In this sense, Ada's
+type system  encourages programmers to think about data at a high level of
+abstraction. Ada supports overlays and unchecked conversions as a way of
+converting between unrelated data type, which are typically used for
+interfacing with low-level elements such as registers.
+
+In Ada, arrays aren't interchangeable with operations on pointers like in C.
+Also, array types are considered first-class citizens and have dedicated
+semantics such as the availability of the array's boundaries at run-time.
+Therefore, unhandled array overflows are impossible unless checks are
+suppressed. Any discrete type can serve as an array index, and we can specify
+both the starting and ending bounds. In addition, Ada offers high-level
+operations for copying, slicing, and assigning values to arrays.
+
+Although Ada supports pointers, most situation that would require a pointer in
+C do not in Ada. In the vast majority of the cases, indirect memory management
+can be hidden from the developer and thus prevent many potential errors. In
+C, pointers are typically used to pass references to subprograms, for example.
+In contrast, Ada supports output parameters (:ada:`out` mode), as well as
+:ada:`in out` parameters.
+
+When translating pointers from C code to Ada, we need to assess whether they
+are needed in the first place. Ada pointers (access types) should only be used
+with complex structures that cannot be allocated at run-time. There are many
+situations that would require a pointer in C, but do not in Ada. For example,
+arrays |mdash| even when dynamically allocated |mdash|, results of functions,
+passing of large structures as parameters, access to registers, etc.
+
+Because of the absence of namespaces, global names in C tend to be very long.
+Also, because of the absence of overloading, they can even encode type names in
+their name. In Ada, a package is a namespace. Also, we can use packages to
+declare private types, which are useful for preventing users from directly
```

Comment:
I would argue that private types are the the fundamental artifact of Ada, in much the same way that the class construct is in C++ or Java. Therefore I think this needs more emphasis.

Reply:
Ok, I've just added two use-cases of private types...